src/mainboard/starfive/visionfive2/bt0/src/init.rs

```diff
@@ -0,0 +1,677 @@
+/*
+ * const GPIO_OUT_FUNC_OFF: u8 = 0x00;
+ * const GPIO_OUT_FUNC_ON: u8 = 0x01;
+ * const GPIO_OUT_FUNC_XXX: u8 = 0x02;
+ * const GPIO_OUT_FUNC_XXX: u8 = 0x03;
+ * const GPIO_OUT_FUNC_XXX: u8 = 0x04;
+ * const GPIO_OUT_FUNC_XXX: u8 = 0x05;
+ * const GPIO_OUT_FUNC_XXX: u8 = 0x06;
+ * const GPIO_OUT_FUNC_XXX: u8 = 0x07;
+ * const GPIO_OUT_FUNC_XXX: u8 = 0x08;
+ * const GPIO_OUT_FUNC_XXX: u8 = 0x09;
+ * const GPIO_OUT_FUNC_XXX: u8 = 0x0a;
+ * const GPIO_OUT_FUNC_XXX: u8 = 0x0b;
+ * const GPIO_OUT_FUNC_XXX: u8 = 0x0c;
+ * const GPIO_OUT_FUNC_XXX: u8 = 0x0d;
+ * const GPIO_OUT_FUNC_XXX: u8 = 0x0e;
+ * const GPIO_OUT_FUNC_XXX: u8 = 0x0f;
+ * const GPIO_OUT_FUNC_XXX: u8 = 0x10;
+ * const GPIO_OUT_FUNC_XXX: u8 = 0x11;
+ * const GPIO_OUT_FUNC_XXX: u8 = 0x12;
+ * const GPIO_OUT_FUNC_XXX: u8 = 0x13;
+ * const GPIO_OUT_FUNC_XXX: u8 = 0x14;
+ * const GPIO_OUT_FUNC_XXX: u8 = 0x15;
+ * const GPIO_OUT_FUNC_XXX: u8 = 0x16;
+ * const GPIO_OUT_FUNC_XXX: u8 = 0x17;
+ * const GPIO_OUT_FUNC_XXX: u8 = 0x18;
+ * const GPIO_OUT_FUNC_XXX: u8 = 0x19;
+ * const GPIO_OUT_FUNC_XXX: u8 = 0x1a;
+ * const GPIO_OUT_FUNC_XXX: u8 = 0x1b;
+ * const GPIO_OUT_FUNC_XXX: u8 = 0x1c;
+ * const GPIO_OUT_FUNC_XXX: u8 = 0x1d;
+ * const GPIO_OUT_FUNC_XXX: u8 = 0x1e;
+ * const GPIO_OUT_FUNC_XXX: u8 = 0x1f;
+ * const GPIO_OUT_FUNC_XXX: u8 = 0x20;
+ * const GPIO_OUT_FUNC_XXX: u8 = 0x21;
+ * const GPIO_OUT_FUNC_XXX: u8 = 0x22;
+ * const GPIO_OUT_FUNC_XXX: u8 = 0x23;
+ * const GPIO_OUT_FUNC_XXX: u8 = 0x24;
+ * const GPIO_OUT_FUNC_XXX: u8 = 0x25;
+ * const GPIO_OUT_FUNC_XXX: u8 = 0x26;
+ * const GPIO_OUT_FUNC_XXX: u8 = 0x27;
+ * const GPIO_OUT_FUNC_XXX: u8 = 0x28;
+ * const GPIO_OUT_FUNC_XXX: u8 = 0x29;
+ * const GPIO_OUT_FUNC_XXX: u8 = 0x2a;
+ * const GPIO_OUT_FUNC_XXX: u8 = 0x2b;
+ * const GPIO_OUT_FUNC_XXX: u8 = 0x2c;
+ * const GPIO_OUT_FUNC_XXX: u8 = 0x2d;
+ * const GPIO_OUT_FUNC_XXX: u8 = 0x2e;
+ * const GPIO_OUT_FUNC_XXX: u8 = 0x2f;
+ * const GPIO_OUT_FUNC_XXX: u8 = 0x30;
+ * const GPIO_OUT_FUNC_XXX: u8 = 0x31;
+ * NOTE: GPIO OUT has 49 functions
+ */
+
+/*
+pub const SYS_SYSCON_BASE: usize = 0x1303_0000;
+const SYS_SYSCON_00: usize = SYS_SYSCON_BASE;
+const SYS_SYSCON_04: usize = SYS_SYSCON_BASE + 0x0004;
+const SYS_SYSCON_08: usize = SYS_SYSCON_BASE + 0x0008;
+const SYS_SYSCON_12: usize = SYS_SYSCON_BASE + 0x000c;
+const SYS_SYSCON_16: usize = SYS_SYSCON_BASE + 0x0010;
+const SYS_SYSCON_20: usize = SYS_SYSCON_BASE + 0x0014;
+const SYS_SYSCON_24: usize = SYS_SYSCON_BASE + 0x0018;
+const SYS_SYSCON_28: usize = SYS_SYSCON_BASE + 0x001C;
+const SYS_SYSCON_32: usize = SYS_SYSCON_BASE + 0x0020;
+const SYS_SYSCON_36: usize = SYS_SYSCON_BASE + 0x0024;
+
+// TODO: Figure out clocks...
+const CLK_ENABLE_MASK: u32 = 0x8000_0000;
+*/
+
+/* BELOW CODE IS FROM JH7100 */
+
+fn write_32(reg: u32, val: u32) {
+    unsafe {
+        write_volatile(reg as *mut u32, val);
+    }
+}
+
+fn read_32(reg: u32) -> u32 {
+    unsafe { read_volatile(reg as *mut u32) }
+}
+
+pub const CLKGEN_BASE: u32 = 0x1180_0000;
```

Comment:
```suggestion
pub const CLKGEN_BASE: u32 = 0x1302_0000;
```

https://doc-en.rvspace.org/JH7110/TRM/JH7110_TRM/system_memory_map.html
https://github.com/starfive-tech/u-boot/blob/1539c1fb5a498fd7d20c2cffc84c187d703c8b66/arch/riscv/dts/jh7110.dtsi#L268

Reply:
I know. This is copied from the previous SoC and only here for reference. We are not at the point where we deal with clocks. The current work is around issues with print macros.

Reply:
Ok, just following along, and noticed this bit. If it's more helpful, I can hold off on review until a later point.

Reply:
We went through this a bit in today's live stream. I have defined a new constant `SYS_CRG_BASE`, aligning with the docs. DDR init is the next step now; printing works fine after fixing the SRAM address in the linker script.